compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2553,4 +2583,182 @@ public int hashCode() {
+        BLangExpression firstElement = (BLangExpression) iterator.next();
+        BType firstElementType = firstElement.type;
+        String defaultFillValue = getDefaultFillValue(firstElement);
+        if (defaultFillValue.equals(DefaultValues.UNKNOWN.getValue())) {
+            return false;
+        }
+        if (firstElement.toString().equals(defaultFillValue)) {
+            defaultFillValuePresent = true;
+        }
+
+        while (iterator.hasNext()) {
+            Object value = iterator.next();
+            BType valueType = ((BLangExpression) value).type;
+            if (!isSameType(valueType, firstElementType)) {
+                return false;
+            }
+            if (!defaultFillValuePresent && value.toString().equals(defaultFillValue)) {
+                defaultFillValuePresent = true;
+            }
+        }
+        return defaultFillValuePresent;
+    }
+
+    private String getDefaultFillValue(BType type) {
+        switch(type.getKind()) {
+            case INT:
+            case BYTE:
+                return DefaultValues.INTEGER.getValue();
+            case STRING:
+                return DefaultValues.STRING.getValue();
+            case DECIMAL:
+            case FLOAT:
+                return DefaultValues.FLOAT.getValue();
+            case BOOLEAN:
+                return DefaultValues.BOOLEAN.getValue();
+            case NIL:
+                return DefaultValues.NIL.getValue();
+            case ERROR:
+            case TYPEDESC:
+                return DefaultValues.UNKNOWN.getValue();
+            default:
+                if (type instanceof BFiniteType) {
+                    return getDefaultFillValue((BFiniteType) type);
+                }
+                return DefaultValues.UNKNOWN.getValue();
+        }
+    }
+
+    private boolean checkFillerValue(BRecordType type) {
+        for (BField field : type.fields) {
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.OPTIONAL)) {
+                continue;
+            }
+            if (Symbols.isFlagOn(field.symbol.flags, Flags.REQUIRED)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean checkFillerValue(BArrayType type) {
+        if (type.size == -1) {
+            return true;
+        }
+        return hasFillerValue(type.eType);
+    }
+
+    private String getDefaultFillValue(BLangExpression expr) {
+        return getDefaultFillValue(expr.type);
+    }
+
+    private String getDefaultFillValue(BFiniteType finiteType) {
+        if (finiteType.getValueSpace().size() == 1) {
+            for (BLangExpression valueLiteral : finiteType.getValueSpace()) {
+                if (!(valueLiteral instanceof BLangLiteral)) {
+                    return DefaultValues.UNKNOWN.getValue();
+                }
+                return valueLiteral.toString();
+            }
+        }
+        return DefaultValues.UNKNOWN.getValue();
```

Comment:
What about when the value space has more than one value but are all of the same type?